### PR TITLE
Fix bugs

### PR DESCRIPTION
**Pull request by Artlands**
1. Fixed RevMem bugs that cannot write/read data across more than two pages; 
2. Fixed bugs in `RevProc::DecodeInst`, where slli, srli, srai, etc. and the imm of the B instruction are not decoded correctly. 
3. Updated RV32I and RV64I instructions implementations; reorganized the instruction tables for easy debugging.

### CODE REVIEW

**Comment by jleidel** [comment]
@Artlands, note the major changes we just merged to devel.  The entire RevMem interface has been revamped in order to support memHierarchy.  The old interfaces are still present, but are marked with `deprecated` attributes.  You'll need to rebase your branch.

**Comment by Artlands** [comment]
Rebased the fixbugs branch and added several other bug fixings.

File: src/RevMem.cc [hunk 1]
```diff
@@ -372,8 +372,9 @@ bool RevMem::ReadMem(uint64_t Addr, size_t Len, void *Target,
 
       // Read data
       BaseMem = &physMem[tmpPhysAddr];
-      ctrl->sendREADRequest(Addr, (uint64_t)(BaseMem), tmpLen, Target, flags);
-
+      ctrl->sendREADRequest(Addr, (uint64_t)(BaseMem), tmpLen, DataMem, flags);
+      DataMem = DataMem + lenCount;
```
Comment (jleidel) [inline src/RevMem.cc:376]: why are you incrementing DataMem

File: src/RevMem.cc [hunk 2]
```diff
@@ -372,8 +372,9 @@ bool RevMem::ReadMem(uint64_t Addr, size_t Len, void *Target,
 
       // Read data
       BaseMem = &physMem[tmpPhysAddr];
-      ctrl->sendREADRequest(Addr, (uint64_t)(BaseMem), tmpLen, Target, flags);
-
+      ctrl->sendREADRequest(Addr, (uint64_t)(BaseMem), tmpLen, DataMem, flags);
```
Comment (jleidel) [inline src/RevMem.cc:375]: this is redundant

File: src/RevMem.cc [hunk 3]
```diff
@@ -382,7 +383,7 @@ bool RevMem::ReadMem(uint64_t Addr, size_t Len, void *Target,
     std::cout << "Warning: Reading off end of page... " << std::endl;
 #endif
   }else{
-    ctrl->sendREADRequest(Addr, (uint64_t)(BaseMem), Len, Target, flags);
+    ctrl->sendREADRequest(Addr, (uint64_t)(BaseMem), Len, DataMem, flags);
```
Comment (jleidel) [inline src/RevMem.cc:386]: redundant

File: src/RevProc.cc
Comment (jleidel) [inline src/RevProc.cc:1]: Split all xBGAS commits into a separate PR

File: src/RevProc.cc [hunk 4]
```diff
@@ -956,309 +956,308 @@ RevInst RevProc::DecodeCompressed(uint32_t Inst){
 }
 
 RevInst RevProc::DecodeRInst(uint32_t Inst, unsigned Entry){
-  RevInst DInst;
+  RevInst RInst;
```
Comment (jleidel) [inline src/RevProc.cc:955]: don't rename things unnecessarily

File: src/RevProc.cc [hunk 5]
```diff
@@ -956,309 +956,308 @@ RevInst RevProc::DecodeCompressed(uint32_t Inst){
...
 
   // encodings
-  DInst.opcode  = InstTable[Entry].opcode;
-  DInst.funct3  = InstTable[Entry].funct3;
-  DInst.funct2  = 0x0;
-  DInst.funct7  = 0x0;
+  IInst.opcode  = InstTable[Entry].opcode;
+  IInst.funct3  = InstTable[Entry].funct3;
+  IInst.funct2  = 0x0;
+  IInst.funct7  = InstTable[Entry].funct7; // For ecall, ebreak, slli, srli, srai
```
Comment (jleidel) [inline src/RevProc.cc:1018]: this is wrong

File: include/insns/RV32I.h [hunk 6]
```diff
@@ -434,15 +467,19 @@ namespace SST{
...
 
       static bool lh(RevFeature *F, RevRegFile *R,RevMem *M,RevInst Inst) {
+        uint32_t Tmp32;
+        uint64_t Tmp64;
         if( F->IsRV32() ){
-          //SEXT(R->RV32[Inst.rd],M->ReadU16( (uint64_t)(R->RV32[Inst.rs1]+(int32_t)(td_u32(Inst.imm,12)))),32);
-          M->ReadVal((uint64_t)(R->RV32[Inst.rs1]+(int32_t)(td_u32(Inst.imm,12))),
+          SEXT(Tmp32, Inst.imm, 12);
+          // SEXT(R->RV32[Inst.rd], M->ReadU16((uint64_t)(R->RV32[Inst.rs1] + Tmp32)), 16);
+          M->ReadVal((uint64_t)(R->RV32[Inst.rs1] + Tmp32),
```
Comment (jleidel) [inline include/insns/RV32I.h:475]: this is broken

File: include/insns/RV32I.h [hunk 7]
```diff
@@ -911,63 +981,63 @@ namespace SST{
       //            <rs2Class> <rs3Class> <format> <func> <nullEntry>
       // ----------------------------------------------------------------------
       std::vector<RevInstEntry> RV32ITable = {
-      {RevInstEntryBuilder<RevInstDefaults>().SetMnemonic("lui %rd, $imm"  ).SetCost(1).SetOpcode(0b0110111).SetFunct3(0b0).SetFunct7(0b0).SetrdClass(RegGPR).Setrs1Class(RegUNKNOWN).Setrs2Class(RegUNKNOWN).Setrs3Class(RegUNKNOWN).Setimm12(0b0).Setimm(FUnk).SetFormat(RVTypeU).SetImplFunc(&lui ).InstEntry},
```
Comment (jleidel) [inline include/insns/RV32I.h:914]: why are you rewriting the entire instruction table.  this is unnecessary

**Comment by jleidel** [comment]
There are a large number of issues with your PR.  You're unnecessarily rewriting a tremendous amount of code that doesn't need to be changed.  Don't change variable names simply because you don't like the naming conventions.  It makes reviewing PR's extremely time consuming

File: include/insns/RV32I.h (hunk 7 above)
Comment (Artlands) [reply include/insns/RV32I.h:914]: I reformated the instruction table for easy debugging. I noticed that in the previous instruction table, it is hard to check functX fields and immediate fields with the RISC-V specifications.

File: src/RevProc.cc (hunk 5 above)
Comment (Artlands) [reply src/RevProc.cc:1018]: > this is wrong

Setting funct 7 to 0 will not be able to differentiate `srli` from `srai`.

File: src/RevMem.cc (hunk 1 above)
Comment (Artlands) [reply src/RevMem.cc:376]: > why are you incrementing DataMem

The memory read/write operation is implemented on a page basis. Increasing DataMem is to move to the target on the next page.

File: include/insns/RV32I.h [hunk 8]
```diff
@@ -251,75 +251,85 @@ namespace SST{
...
+        uint32_t Tmp = Inst.imm << 12;
         if( F->IsRV32() ){
-          R->RV32[Inst.rd] = 0x00;
-          R->RV32[Inst.rd] = (Inst.imm << 12);
+          R->RV32[Inst.rd] = Tmp;
           R->RV32_PC += Inst.instSize;
         }else{
-          R->RV64[Inst.rd] = 0x00;
-          R->RV64[Inst.rd] = (Inst.imm << 12);
+          SEXT(R->RV64[Inst.rd], Tmp, 32);
```
Comment (jleidel) [inline include/insns/RV32I.h:259]: this is incorrect.  the spec says that LUI loads unsigned immediates into the top 20 bits of the target.  
```
LUI (load upper immediate) is used to build 32-bit constants and uses the U-type format. LUI places the U-immediate value in the top 20 bits of the destination register rd, filling in the lowest 12 bits with zeros.
```

File: include/insns/RV64I.h [hunk 9]
```diff
@@ -85,18 +85,23 @@ namespace SST{
```
Comment (donofrio) [inline include/insns/RV64I.h:85]: Changes from this PR that addressed bugs in RV64I instructions have been pushed to devel in commit b037382. Changes that just re-structured code are not included in that commit

File: src/RevProc.cc
Comment (donofrio) [reply src/RevProc.cc:1]: Changes to the instruction decoder logic will require tests to demonstrate correct functionality (and how the previous code was functioning incorrectly) as this is fairly fundamental logic to the simulator
